### PR TITLE
Parse and emit `array.len` without a type annotation

### PR DESCRIPTION
**Pull request by tlively**
Test that we can still parse the old annotated form as well.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5148** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5148" target="_blank">Graphite</a> 
    * **PR #5151** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5151" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/5151?utm_source=stack-comment).

File: src/wasm/wasm-s-parser.cpp [hunk 1]
```diff
@@ -3027,9 +3027,14 @@ Expression* SExpressionWasmBuilder::makeArraySet(Element& s) {
 }
 
 Expression* SExpressionWasmBuilder::makeArrayLen(Element& s) {
-  // Ignore the type annotation and don't bother validating it.
-  parseHeapType(*s[1]);
-  auto ref = parseExpression(*s[2]);
+  // There may or may not be a type annotation.
+  Index childIdx = 1;
```
Comment (kripken) [inline src/wasm/wasm-s-parser.cpp:3031]: We have lots of places in this file that use `i` for such a moving index. But I don't feel strongly, maybe this is clearer.

File: src/wasm/wasm-stack.cpp [hunk 2]
```diff
@@ -2169,8 +2169,6 @@ void BinaryInstWriter::visitArraySet(ArraySet* curr) {
 
 void BinaryInstWriter::visitArrayLen(ArrayLen* curr) {
   o << int8_t(BinaryConsts::GCPrefix) << U32LEB(BinaryConsts::ArrayLen);
-  // Unused type index.
-  o << U32LEB(0);
```
Comment (kripken) [inline src/wasm/wasm-stack.cpp:2173]: Shouldn't we emit the annotated version for now until V8 supports both? Or does it already?

**Review by kripken — APPROVED** [review]
lgtm % comment

File: src/wasm/wasm-stack.cpp (hunk 2 above)
Comment (tlively) [reply src/wasm/wasm-stack.cpp:2173]: This is still emitting the annotated version, but the annotation is meaningless. V8 has supported both and ignored the annotation for a while now.

**Comment by tlively** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/5151?utm_source=stack-comment).


Job Reference: TH3xuFC7pQ278NbtWS9g

**Pull request merged**